staging/src/k8s.io/apiserver/pkg/audit/request_test.go

```diff
@@ -54,3 +65,238 @@ func TestMaybeTruncateUserAgent(t *testing.T) {
+		obj         runtime.Object
+		objGV       schema.GroupVersion
+		sourceGVR   schema.GroupVersionResource
+		subResource string
+		expectEvent *auditinternal.Event
+	}{
+		{
+			desc:        "auditEvent is nil",
+			auditEvent:  nil,
+			expectEvent: nil,
+		},
+		{
+			desc: "encodeObject error",
+			auditEvent: &auditinternal.Event{
+				Level:     auditinternal.LevelRequest,
+				ObjectRef: nil,
+			},
+			obj: &authnv1.TokenReview{
+				Spec: authnv1.TokenReviewSpec{
+					Token:     "foo",
+					Audiences: []string{"k8s", "api"},
+				},
+			},
+			objGV:     schema.GroupVersion{Group: "authentication.k8s.io", Version: "v2"},
+			sourceGVR: schema.GroupVersionResource{Group: "authentication.k8s.io", Version: "v1", Resource: "tokenReviews"},
+			expectEvent: &auditinternal.Event{
+				Level: auditinternal.LevelMetadata,
+				ObjectRef: &auditinternal.ObjectReference{
+					Resource:   "tokenReviews",
+					APIGroup:   "authentication.k8s.io",
+					APIVersion: "v1",
+				},
+			},
+		},
+		{
+			desc: "log resource at LevelMetadata",
+			auditEvent: &auditinternal.Event{
+				Level:     auditinternal.LevelMetadata,
+				ObjectRef: nil,
+			},
+			obj: &authnv1.TokenReview{
+				Spec: authnv1.TokenReviewSpec{
+					Token:     "foo",
+					Audiences: []string{"k8s", "api"},
+				},
+			},
+			objGV:     schema.GroupVersion{Group: "authentication.k8s.io", Version: "v1"},
+			sourceGVR: schema.GroupVersionResource{Group: "authentication.k8s.io", Version: "v1", Resource: "tokenReviews"},
+			expectEvent: &auditinternal.Event{
+				Level: auditinternal.LevelMetadata,
+				ObjectRef: &auditinternal.ObjectReference{
+					Resource:   "tokenReviews",
+					APIGroup:   "authentication.k8s.io",
+					APIVersion: "v1",
+				},
+			},
+		},
+		{
+			desc: "log cross-group subResource at LevelRequest",
+			auditEvent: &auditinternal.Event{
+				Level: auditinternal.LevelRequest,
+				// gvr is not in ObjectRef
+				ObjectRef: nil,
+			},
+			obj: &authnv1.TokenRequest{
+				Spec: authnv1.TokenRequestSpec{
+					ExpirationSeconds: &expTime,
+				},
+				Status: authnv1.TokenRequestStatus{
+					Token:               "foo",
+					ExpirationTimestamp: metav1.Time{Time: time.Now().Add(3600 * time.Second)},
+				},
+			},
+			objGV:       schema.GroupVersion{Group: "authentication.k8s.io", Version: "v1"},
+			sourceGVR:   schema.GroupVersionResource{Group: "", Version: "v1", Resource: "serviceaccounts"},
+			subResource: "token",
+			expectEvent: &auditinternal.Event{
+				Level: auditinternal.LevelRequest,
+				ObjectRef: &auditinternal.ObjectReference{
+					Resource:    "serviceaccounts",
+					APIGroup:    "",
+					APIVersion:  "v1",
+					Subresource: "token",
+				},
+			},
+		},
+	}
+	for _, tc := range testCases {
+		t.Run(tc.desc, func(t *testing.T) {
+			LogRequestObject(tc.auditEvent, tc.obj, tc.objGV, tc.sourceGVR, tc.subResource, scheme.Codecs)
+
+			// encodeObject error
+			if tc.desc == "encodeObject error" {
```

Comment:
Are such IF-Conditions necessary？ Moving into universal Expected statement maybe more grace

Reply:
em... 
I think I need to implement return errors to write ut well, otherwise I cannot avoid if-else, because I cannot build the object returned when LEVELREQUEST is related, it is related to the current time.